### PR TITLE
Fix bugs and improve pxr ##print

### PR DESCRIPTION
**Pull request by trufae**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**

pxr is bind to asm.bits, theres no way to use a different word size until...NOW

also i fixed a boundary bug that was shown in one BROKEN test.

there are still more formats to be supported as output for pxr, but i think this wont come until i finish and get merged the table PR, so for me its fine to merge this as is

**Test plan**

i wrote tests

**Closing issues**

no issue listed

### CODE REVIEW

File: libr/core/cmd_print.c [hunk 1]
```diff
@@ -6168,63 +6235,26 @@ l = use_blocksize;
...
-					core->print->flags &= ~R_PRINT_FLAGS_REFS;
-					core->print->cols = ocols;
+				int mode = input[2];
+				int wordsize = core->anal->bits / 8;
+				if (mode == '?') {
+					eprintf ("Usage: pxr[1248][jq] [length]\n");
+					break;
+				}
+				if (isdigit (mode)) {
+					wordsize = atoi (input + 2);
```
Comment (ret2libc) [inline libr/core/cmd_print.c:6245]: I think you should do something like:
```
char tmp[2] = {input[2], 0};
wordsize = atoi (tmp);
```
just to be sure there aren't other numbers after input[2]

File: libr/util/print.c [hunk 2]
```diff
@@ -1263,7 +1269,7 @@ R_API void r_print_hexdump(RPrint *p, ut64 addr, const ut8 *buf, int len, int ba
 			bool eol = false;
 			if (!eol && p && p->flags & R_PRINT_FLAGS_REFS) {
 				ut64 off = 0;
-				if (i + 8 < len) {
+				if (i + 8 <= len) {
```
Comment (ret2libc) [inline libr/util/print.c:1272]: this looks wrong. The for-loop does `i < len`, so I think there shouldn't be `<=` here.

File: test/db/cmd/cmd_px [hunk 3]
```diff
@@ -108,16 +108,113 @@ EOF
...
+0x00600aec ..[ null bytes ]..   00000000 
+0x00600af0 0x0040057e  ~.@. @reloc.__libc_start_main 4195710 (.plt) R X 'push 2'
+0x00600af4 ..[ null bytes ]..   00000000 
+0x00600af8 0x0040058e  ..@. @reloc.fgets 4195726 (.plt) R X 'push 3'
+0x00600afc ..[ null bytes ]..   00000000 
+0x00600ae0 0x000000000040055e   ^.@..... @reloc.printf 4195678 (.plt) R X 'push 0'
+0x00600ae8 0x000000000040056e   n.@..... @reloc.puts 4195694 (.plt) R X 'push 1'
+0x00600af0 0x000000000040057e   ~.@..... @reloc.__libc_start_main 4195710 (.plt) R X 'push 2'
+0x00600af8 0x000000000040058e   ..@..... @reloc.fgets 4195726 (.plt) R X 'push 3'
+5e05  ^. @reloc.printf 4195678 (.plt) R X 'push 0'
```
Comment (ret2libc) [inline test/db/cmd/cmd_px:154]: the output of `pxr1` and `pxr2` looks completely different from `pxr4`/`pxr8`.
There should be the offset and the hex value value should have `0x` and respect bigendian.

File: test/db/cmd/cmd_px [hunk 4]
```diff
@@ -108,16 +108,113 @@ EOF
...
+
+NAME=pxr1248
+FILE=bins/elf/crackme
+CMDS=<<EOF
+pxr4 32@0x00600ae0
+pxr8 32@0x00600ae0
+pxr2 32@0x00600ae0
+e asm.bits=32
+pxr 32@0x00600ae0
+pxr2j 16@0x00600ae0~{}
```
Comment (ret2libc) [inline test/db/cmd/cmd_px:139]: If I do `pxr2j` without setting `e asm.bits=32` I get wrong results. Why is that? I would expect to see 16bits-wide values.

**Review by ret2libc — CHANGES_REQUESTED** [review]

File: libr/util/print.c (hunk 2 above)
Comment (trufae) [reply libr/util/print.c:1272]: this is fixing the broken test

File: test/db/cmd/cmd_px (hunk 3 above)
Comment (trufae) [reply test/db/cmd/cmd_px]: fixed

File: test/db/cmd/cmd_px (hunk 4 above)
Comment (trufae) [reply test/db/cmd/cmd_px:139]: should be fixed now

File: libr/util/print.c (hunk 2 above)
Comment (ret2libc) [reply libr/util/print.c:1272]: It looks wrong anyway and I guess it will be detect by ASAN if you try the test with it.

File: libr/include/r_util/r_str.h [hunk 5]
```diff
@@ -150,6 +150,7 @@ R_API char *r_str_path_escape(const char *path);
 R_API int r_str_unescape(char *buf);
 R_API char *r_str_escape(const char *buf);
 R_API char *r_str_escape_sh(const char *buf);
+R_API char *r_str_escape_r2(const char *buf);
```
Comment (ret2libc) [inline libr/include/r_util/r_str.h:153]: this is out of scope for this PR and considering the amount of r2-shell injections we have, I don't think this single escape will "save the world". I would much rather have this in a separate PR with *a lot* of unit tests to make sure it works as intended.

File: libr/core/cmd_print.c [hunk 6]
```diff
@@ -6168,63 +6273,31 @@ l = use_blocksize;
...
+					eprintf ("Usage: pxr[1248][*,jq] [length]\n");
+					break;
+				}
+				if (mode && isdigit (mode)) {
+					char tmp[2] = {input[2], 0};
+					wordsize = atoi (tmp);
+					mode = input[3];
+				}
+				if (!mode) {
+					mode = 1;
```
Comment (ret2libc) [inline libr/core/cmd_print.c:6288]: What is `mode == 1`?

File: libr/core/cmd_print.c [hunk 7]
```diff
@@ -4486,6 +4486,111 @@ static void r_core_disasm_table(RCore * core, int l, const char *input) {
...
+						char *ns = r_str_escape (rstr);
+						pj_ks (pj, "ref", r_str_trim_head_ro (ns));
+						pj_end (pj);
+						free (ns);
+					}
+					withref = 1;
+				}
+				refs = rstr;
+			}
+			if (mode == '*' && refs && *refs) {
```
Comment (ret2libc) [inline libr/core/cmd_print.c:4544]:
```suggestion
			if (mode == '*' && R_STR_ISNOTEMPTY (refs)) {
```

File: libr/core/cmd_print.c [hunk 8]
```diff
@@ -4486,6 +4486,111 @@ static void r_core_disasm_table(RCore * core, int l, const char *input) {
...
+				// Show only the mapped ones?
+				char *rs = r_str_escape_r2 (refs);
+				r_cons_printf ("f pxr.%"PFMT64x"=0x%"PFMT64x" # %s\n", val, addr, rs);
+				free (rs);
+			}
+			if (mode == 'q' && refs && *refs) {
```
Comment (ret2libc) [inline libr/core/cmd_print.c:4550]:
```suggestion
			} else if (mode == 'q' && R_STR_ISNOTEMPTY (refs)) {
```

File: libr/core/cmd_print.c [hunk 9]
```diff
@@ -4486,6 +4486,111 @@ static void r_core_disasm_table(RCore * core, int l, const char *input) {
...
+				pj_kn (pj, "addr", addr);
+				pj_kn (pj, "value", val);
+			}
+
+			// XXX: this only works in little endian
+			withref = 0;
+			char *refs = NULL;
+			if (core->print->hasrefs) {
+				char *rstr = core->print->hasrefs (core->print->user, val, true);
+				if (rstr && *rstr) {
```
Comment (ret2libc) [inline libr/core/cmd_print.c:4532]:
```suggestion
				if (R_STR_ISNOTEMPTY (rstr)) {
```

File: libr/core/cmd_print.c [hunk 10]
```diff
@@ -4486,6 +4486,111 @@ static void r_core_disasm_table(RCore * core, int l, const char *input) {
...
+				char *rs = r_str_escape_r2 (refs);
+				r_cons_printf ("f pxr.%"PFMT64x"=0x%"PFMT64x" # %s\n", val, addr, rs);
+				free (rs);
+			}
+			if (mode == 'q' && refs && *refs) {
+				r_cons_printf ("%s\n", refs);
+			}
+			if (t) {
+				r_table_add_rowf (t, "xxs", addr, val, refs);
+				refs = NULL;
```
Comment (ret2libc) [inline libr/core/cmd_print.c:4555]: doesn't `refs` leak this way?

Comment (trufae) [reply libr/core/cmd_print.c:4555]: nope, rowf owns the pointer passed

**Comment by trufae** [comment]
in bsd it fails in here wtf
Screenshot 2020-09-24 at 17 29 22

File: libr/include/r_util/r_str.h (hunk 5 above)
Comment (trufae) [reply libr/include/r_util/r_str.h:153]: its needed for pxr*, its not really escaping, just filtering dangerous chars. i can do a separate PR for this indeed

File: libr/core/cmd_print.c (hunk 6 above)
Comment (trufae) [reply libr/core/cmd_print.c:6289]: something that is different than 0 and different from other defined modes.

File: libr/core/cmd_print.c (hunk 10 above)
Comment (ret2libc) [reply libr/core/cmd_print.c:4552]: I don't think so. `r_table_add_rowf` does a strcpy AFAIS.

File: libr/core/cmd_print.c (hunk 6 above)
Comment (ret2libc) [reply libr/core/cmd_print.c:6286]: and what does that mean exactly? :/ what mode is that?

File: libr/core/cmd_print.c (hunk 10 above)
Comment (trufae) [reply libr/core/cmd_print.c:4552]: indeed. dunno what was thinking

File: libr/util/print.c (hunk 2 above)
Comment (trufae) [reply libr/util/print.c]: asan is not complaining here, but it was complaining in other places. ive added more tests and fixed the bugs i spotted by undefiuned behaviour and out of bounds accesses. current code is safer, but still needs more love. as longas i have two prs for pxr i would love to get them merged before going further to avoid more conflicts between them

Comment (trufae) [reply libr/util/print.c]: fixed

**Review by XVilka — CHANGES_REQUESTED** [review]
3 failing tests:
```
[XX] C:\projects\radare2\test\db\cmd\cmd_pxw pxw bytes < 4
ANSICON=1 R2_NOPLUGINS=1 radare2 -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'e hex.cols=8
pxw 2~[1]
' bins/pe/hello-mingw32
-- stdout
@@ -0,0 +1 @@
+0x00008955
[XX] C:\projects\radare2\test\db\cmd\cmd_px pxr1248
ANSICON=1 R2_NOPLUGINS=1 radare2 -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'pxr4 32@0x00600ae0
pxr8 32@0x00600ae0
pxr2 32@0x00600ae0
e asm.bits=32
pxr 32@0x00600ae0
pxr2j 16@0x00600ae0~{}
' bins/elf/crackme
-- stdout
@@ -9,20 +9,20 @@
 0x00600ae0 0x000000000040055e   ^.@..... @reloc.printf 4195678 (.plt) R X 'push 0'
 0x00600ae8 0x000000000040056e   n.@..... @reloc.puts 4195694 (.plt) R X 'push 1'
 0x00600af0 0x000000000040057e   ~.@..... @reloc.__libc_start_main 4195710 (.plt) R X 'push 2'
-0x00600af8 0x000000000040058e   ..@..... @reloc.fgets
-0x00600ae0 5e05  ^. @reloc.printf
+0x00600af8 0x000000000040058e   ..@.....
+0x00600ae0 5e05  ^.
 0x00600ae2 4000  @.
 0x00600ae4 0000  ..
 0x00600ae6 0000  ..
-0x00600ae8 6e05  n. @reloc.puts
+0x00600ae8 6e05  n.
 0x00600aea 4000  @.
 0x00600aec 0000  ..
 0x00600aee 0000  ..
-0x00600af0 7e05  ~. @reloc.__libc_start_main
+0x00600af0 7e05  ~.
 0x00600af2 4000  @.
 0x00600af4 0000  ..
 0x00600af6 0000  ..
-0x00600af8 8e05  .. @reloc.fgets
+0x00600af8 8e05  ..
 0x00600afa 4000  @.
 0x00600afc 0000  ..
 0x00600afe 0000  ..
[XX] C:\projects\radare2\test\db\cmd\cmd_px pxr
ANSICON=1 R2_NOPLUGINS=1 radare2 -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'pxr 48 @ 0x00600ae0
?e ===
pxr 24 @ 0x00600ae0+32
' bins/elf/crackme
-- stdout
@@ -3,7 +3,7 @@
 0x00600af0 0x000000000040057e   ~.@..... @reloc.__libc_start_main 4195710 (.plt) R X 'push 2'
 0x00600af8 0x000000000040058e   ..@..... @reloc.fgets 4195726 (.plt) R X 'push 3'
 0x00600b00 0x000000000040059e   ..@..... @reloc.strcmp 4195742 (.plt) R X 'push 4'
-0x00600b08 0x00000000004005ae   ..@..... @reloc.fflush
+0x00600b08 0x00000000004005ae   ..@.....
 ===
 0x00600b00 0x000000000040059e   ..@..... @reloc.strcmp 4195742 (.plt) R X 'push 4'
 0x00600b08 0x00000000004005ae   ..@..... @reloc.fflush 4195758 (.plt) R X 'push 5'
```

**Review by XVilka — APPROVED** [review]

**Comment by trufae** [comment]
I will continue my other pxr PR when this gets merged, a month of waiting is probably enough to mature the changes

File: libr/core/cmd_print.c [hunk 11]
```diff
@@ -4486,6 +4486,111 @@ static void r_core_disasm_table(RCore * core, int l, const char *input) {
...
+	}
+	if (mode == 'j' || mode == ',' || mode == '*' || mode == 'q') {
+		size_t i;
+		const int base = core->anal->bits;
+		if (pj) {
+			pj_a (pj);
+		}
+		const ut8 *buf = core->block;
+		int withref = 0;
+		for (i = 0; i + wordsize < core->blocksize; i += wordsize) {
```
Comment (wargio) [inline libr/core/cmd_print.c:4514]: shouldn't this be `i < core->blocksize` ? `i += wordsize` is always executed at each loop before the loop condition.

File: libr/core/cmd_print.c [hunk 12]
```diff
@@ -4486,6 +4486,111 @@ static void r_core_disasm_table(RCore * core, int l, const char *input) {
...
+			ut64 addr = core->offset + i;
+			ut64 val = buf[i];
+			if (i + wordsize >= len) {
```
Comment (wargio) [inline libr/core/cmd_print.c:4517]: same here. why checking before the word size for the next iteration instead of the current one?

File: libr/core/cmd_print.c [hunk 13]
```diff
@@ -4486,6 +4486,111 @@ static void r_core_disasm_table(RCore * core, int l, const char *input) {
...
+				break;
+			}
+			if (base == 32) {
+				val = r_read_le32 (buf + i);
+				val &= UT32_MAX;
+			} else {
+				val = r_read_le64 (buf + i);
+			}
```
Comment (wargio) [inline libr/core/cmd_print.c:4525]: base can be only 32 or 64 bits? and only little endian?

File: libr/core/cmd_print.c (hunk 12 above)
Comment (trufae) [reply libr/core/cmd_print.c:4517]: Because i cant print an 8 byte word if the remaining bytes in the buffer are less than 8. The check can be moved into the for loop statement. so fixed when i push i guess

File: libr/core/cmd_print.c (hunk 11 above)
Comment (trufae) [reply libr/core/cmd_print.c:4514]: i'm trying to avoid parsing incomplete words. so the check is correct imho

File: libr/core/cmd_print.c (hunk 13 above)
Comment (trufae) [reply libr/core/cmd_print.c:4533]: for the 99% of real use cases yes. but i have extended it in a commit

File: libr/core/cmd_print.c (hunk 11 above)
Comment (wargio) [reply libr/core/cmd_print.c]: oh i see the issue. yes it is correct the check.

**Review by wargio — APPROVED** [review]
LGTM
